### DISCUSSION TITLE
Slint UI that can be controlled using 6 buttons (arrows, enter and escape)

### DISCUSSION THREAD

**Discussion by tjallingt** [slint-ui/slint#5550] [Q&A]
Hi,

I would like to know if there is more information (or just peoples ideas) regarding using SLint for a UI that can be controlled using just 4 arrow keys and enter and back.

Slint does have virtual keyboard support, which will help a lot for this project but I was wondering if there are recommendations for managing and moving focus. Ideally it should be possible to control the focus in a more advanced way than the standard accesibility options from slint.

Some way of having a vertical list, horizontal list and a grid in which the focus of items can be controlled?

Would slint be a good fit for such an interface? Should the focus management be implemented in the slint language or would it be recommended to implement it in rust? could we add (custom) primitives to allow more control over focus than slints built-in accesibilty layer?

**Comment by ogoffart** [comment]
Right now Slint can only do tab. So if you'd like to do that, you'd need to implement that keyboard navigation yourself by handling the key up/down/left/right events to change the focus.

**Reply by tjallingt** [reply]
Yea thats what I expected, but I'm not sure what would be the right strategy for implementing this ourselves.

It would be nice if we could wrap this in a reusable component somehow.
Capturing the Key events can be done in slint using the FocusScope like this:

```rust
FocusScope {
    key-pressed(event) => {
        if (event.text == Key.UpArrow) {
            //
        }
        if (event.text == Key.DownArrow) {
            //
        }
        if (event.text == Key.LeftArrow) {
            //
        }
        if (event.text == Key.RightArrow) {
            //
        }
        if (event.text == Key.Return) {
            //
        }
        if (event.text == Key.Escape) {
            //
        }
        accept
    }
}
```

But then we need logic to "find" the component with the right row/column index.
And it would be nice if the components would automatically get their row/column indexes assigned from their parent.

I would like to create custom `FocusVerticalList`, `FocusHorizontalList` and `FocusGrid` components to manage this but its not clear to me that it is possible to create such custom components (or how to, if it can be done).

**Reply by tronical** [reply]
Olivier and I discussed this a bit a distilled https://github.com/slint-ui/slint/issues/6060 out of this. It seems as if API wise a single property and a relatively straight-forward implementation could solve this. What do you think?

**Reply by tjallingt** [reply]
yes that solution sounds perfect for us. Very elegant.

**Reply by bart-sensirion-qm** [reply]
Sounds like a great solution that 'just works'™. With this implementation would you test for edge cases where a focus-able element isn't the receiver of focus from any other element and so can't be navigated to with the arrows or should that be done by us?

**Reply by tronical** [reply]
In essence, this would only allow for navigating to `FocusScope` instances that have `keypad-navigation: true;`. Other elements would not be focusable with the arrow keys, if the navigation is initiated by a `FocusScope`.

**Reply by bart-sensirion-qm** [reply]
That seems logical, but filtered for `FocusScope` elements with `keypad-navigation: true;` the same edge case where one such element has no way to be reached from one of those other elements could still appear (I think).

**Reply by tronical** [reply]
Hmm, could you describe how such a scenario would look like, in terms of placement on the screen?

**Reply by bart-sensirion-qm** [reply]
![focus management possible problem](https://github.com/user-attachments/assets/0c645a0e-8840-4d45-b143-f7adbb0f903c)
Depending on how the next element is determined, the corner elements could move focus to each other which makes the middle element unreachable.

**Reply by tronical** [reply]
Yes indeed. This would require say pressing two keys at the same time (diagonal movement). But I guess what you're saying is that the application should be able to decide if say the bottom right element is focused and the user presses left, if that should or should navigate via the center element or not?

Maybe for such setups it would make sense to allow an optional override? But I see there are also gaps. So for example:

```slint
top-left := FocusScope {
    // ...
    keypad-navigation: true;
}
top-right := FocusScope {
    // ...
    keypad-navigation: true;
}
center := FocusScope {
    // ...
    keypad-navigation: true;
    // What should be the override here? Up or down?
}
bottom-left := FocusScope {
   // ...
    keypad-navigation: true;
    keypad-navigation-right-towards: center; // right should navigate via center
}
bottom-right := FocusScope {
    // ...
    keypad-navigation: true;
    keypad-navigation-left-towards: center; // left should navigate via center
}
```

**Reply by bart-sensirion-qm** [reply]
I don't know if you should fix it that way but I do think that some automated analysis should be done e.g. looking up all the targets for each node on a screen and building a graph to see if all nodes are connected. Maybe even add a viewer for that graph. 

For the fix, maybe just showing a warning for unconnected elements will be enough and people should fix it themselves by changing their UI. Nudging elements or invisibly extending the middle element into a horizontal or vertical line so it will 'naturally' get focus from other elements. Or we could explicitly name the element to go to for each direction to override automatic selection.

I don't think this is an issue that will pop up often as most UI is easily navigable with four directions. It's just that we'd need to make sure it doesn't happen when this is our only way to navigate the page so some automated way to detect it is crucial.
